### DISCUSSION TITLE
Interested on working on a query linter

### DISCUSSION THREAD

**Discussion by kirawi** [SeaQL/summer-of-code#7] [Q&A]
Hello, I am interested on working on [a query linter for SeaORM](https://github.com/SeaQL/summer-of-code/tree/main/2022#4-query-linter-for-seaorm). I'm not familiar with SeaORM, and so I have a few questions as to what the goals of the project would be:

1. `SQLx` advertises "compile-time checked queries" already, and SeaORM seems to depend upon the library. Are you looking to overcome the listed limitations of their checking?
2. Is the objective to create a linter + lints, or just the linter?
3. "The linter should be able to catch syntatic, semantic and logic errors given the schema definition." What kind of semantic and logic errors are you looking to catch? Reading that requirement, I thought of fetching the wrong column from a table, but I'm guessing you mean something different.
4. Are you looking for this to be a CLI tool, or baked into SeaORM?
5. Would you be willing to accept applicants who will be high school graduates, but are taking a gap year due to personal reasons? I'm not sure how well I will stack against other applicants, but I am reasonably confident in my ability to tackle this project.

**Comment by tyt2y3** [comment]
1. Yes. `SQLx` works by connecting to a real database at build time. But I want an approach to be completely decoupled from a database
2. Ideally that would be a linting framework where rules can be added easily in the future, but at least some obvious rules should ship by default to proof the usefulness. It should work by accepting a sea-query AST and schema information
3. Yes and probably some ideas from 
https://www.sea-ql.org/SeaORM/docs/write-test/testing
https://github.com/SeaQL/sea-orm/search?q=lint%3A
4. A cargo feature to be enabled at compile time and some facilities to integrate with cargo test
5. Education / qualification is not a strict requirement (other than complying with GSoC), but we also got to have confidence in your knowledge in computer science so make sure you include relevant information in your application :)

**Reply by kirawi** [reply]
I'm starting to work on a POC, do you have any suggestions for where I can start familiarizing myself with the code that will likely be involved?

**Reply by tyt2y3** [reply]
Um... I really appreciate your eagerness. You can start with "looking and playing around the codebase" and "writing a design document". Do not start writing code yet. We will review the design and implementation approach before really "get hands dirty"

**Reply by tyt2y3** [reply]
I think the SeaQuery AST is a good starting point. And then think about types first e.g. we should not do A + B if the respective columns are not numeric

**Reply by tyt2y3** [reply]
Also, think about what compile-time guarantee Diesel provides and how to gather enough information to do that in SeaORM at runtime (during unit-testing)

**Reply by kirawi** [reply]
Ah I wasn't planning on writing the actual code 😅 By POC I meant figuring out a design that would work. I had a brain fart while trying to use the appropriate word.

**Reply by billy1624** [reply]
Hey @kirawi, you can start by looking into the AST of query builder, https://github.com/SeaQL/sea-query. For example the select statement, https://docs.rs/sea-query/0.24.0/sea_query/query/struct.SelectStatement.html.

**Reply by kirawi** [reply]
For type checking, I was thinking of checking whether the table/column/etc. in a query exists in the database. Is there a method(s) to get a list of the table/column/etc, of the database?

**Reply by billy1624** [reply]
I suggest you look into https://github.com/SeaQL/sea-schema. For different db it was handled differently

**Comment by billy1624** [comment]
Hey @kirawi, GSoC contributor application now opens. Please submit your proposal early so we can review and provide feedbacks :)

**Reply by kirawi** [reply]
I've revised my proposal according to the feedback I received.

**Reply by tyt2y3** [reply]
Got it
